localstack/services/sns/publisher.py

```diff
@@ -0,0 +1,946 @@
+import abc
+import ast
+import base64
+import datetime
+import json
+import logging
+import time
+import traceback
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from typing import Dict, List, Tuple, Union
+
+import requests
+
+from localstack import config
+from localstack.aws.api.lambda_ import InvocationType
+from localstack.aws.api.sns import MessageAttributeMap
+from localstack.aws.api.sqs import MessageBodyAttributeMap
+from localstack.config import external_service_url
+from localstack.services.sns import constants as sns_constants
+from localstack.services.sns.models import (
+    SnsApplicationPlatforms,
+    SnsMessage,
+    SnsStore,
+    SnsSubscription,
+)
+from localstack.utils.aws import aws_stack
+from localstack.utils.aws.arns import (
+    extract_region_from_arn,
+    extract_resource_from_arn,
+    parse_arn,
+    sqs_queue_url_for_arn,
+)
+from localstack.utils.aws.aws_responses import create_sqs_system_attributes
+from localstack.utils.aws.dead_letter_queue import sns_error_to_dead_letter_queue
+from localstack.utils.cloudwatch.cloudwatch_util import store_cloudwatch_logs
+from localstack.utils.json import json_safe
+from localstack.utils.objects import not_none_or
+from localstack.utils.strings import long_uid, md5, to_bytes
+from localstack.utils.time import timestamp_millis
+
+LOG = logging.getLogger(__name__)
+
+# future config flag
+PLATFORM_APPLICATION_REAL = False
+
+
+@dataclass(frozen=True)
+class SnsPublishContext:
+    message: SnsMessage
+    store: SnsStore
+    request_headers: Dict[str, str]
+
+
+@dataclass
```

Comment:
This one's open because we need to reassign `messages` after being filtered.
I guess we can remove the `frozen` from `SnsPublishContext`, it was the one being most instantiated so it was a minor optimisation. I could either make a comment or remove it all together, I don't think it has a lot of value